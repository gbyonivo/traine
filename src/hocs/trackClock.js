import React from 'react';
import PropTypes from 'prop-types';
import dateFns from 'date-fns';
import { getStationAndPosition, extractTime, shouldStartTime } from '../functions';

export default (Component) => {
  class TrackClock extends React.Component {
    constructor() {
      super();
      this.timer = null;
      this.state = {
        currentDateTime: new Date()
      };
      this.addTime = this.addTime.bind(this);
    }

    componentWillMount() {
      this.startTimer();
    }

    componentDidUpdate() {
      this.clearTimer();
      this.startTimer();
    }

    componentWillUnmount() {
      this.clearTimer();
    }

    startTimer() {
      const { stops } = this.props;
      if (!stops || stops.length === 0 || !shouldStartTime(this.props.stops[stops.length - 1], extractTime(this.state.currentDateTime))) {
        this.clearTimer();
        return;
      }
      this.timer = setTimeout(this.addTime, 60000);
    }

    addTime() {
      this.setState(state => ({ currentDateTime: dateFns.addMinutes(state.currentDateTime, 1) }));
    }

    clearTimer() {
      if (this.timer) {
        clearTimeout(this.timer);
      }
    }

    render() {
      return <Component
        {...this.props}
        currentTrainLocation = {getStationAndPosition(this.props.stops, extractTime(this.state.currentDateTime))}
      />;
    }
  }

  TrackClock.defaultProps = {
    stops: []
  };

  TrackClock.propTypes = {
    stops: PropTypes.array.isRequired
  };

  return TrackClock;
};