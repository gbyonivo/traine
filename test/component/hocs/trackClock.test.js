import React from 'react';
import renderer from 'react-test-renderer';
import trackClock from '../../../src/hocs/trackClock';
import { stops } from '../../__testData__';

// REMEMBER TO CHECK IF THERE ARE NO STOPS!!!!!!!

describe('refresh', () => {
  const BaseComponent = ({ prop }) => (<div className={prop} />); // eslint-disable-line
  const TimedComponent = trackClock(BaseComponent);

  it('should return a TimedComponent component when passed a named component', () => {
    const actual = renderer.create(
      <TimedComponent prop="test" stops={stops}/>
    ).toJSON();

    expect(actual).toMatchSnapshot();
  });

  it('should return a TimedComponent with empty stops', () => {
    const actual = renderer.create(
      <TimedComponent prop="test" stops={[]}/>
    ).toJSON();

    expect(actual).toMatchSnapshot();
  });


  it('should refresh every minute as far as time is before last stop', () => {
    jasmine.clock().install();
    jasmine.clock().mockDate(new Date(2018, 7, 21, 1, 0, 0));
    const actual = renderer.create(<TimedComponent prop="test" stops={stops}/>);

    spyOn(TimedComponent.prototype, 'render').and.callThrough();

    jasmine.clock().tick(60000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(1);
    jasmine.clock().tick(60000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(2);
    jasmine.clock().tick(600000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(12);

    jasmine.clock().uninstall();
  });

  it('should not refresh if there are not stops', () => {
    jasmine.clock().install();
    jasmine.clock().mockDate(new Date(2019, 7, 21, 1, 0, 0));
    const actual = renderer.create(<TimedComponent prop="test" stops={[]}/>);

    spyOn(TimedComponent.prototype, 'render').and.callThrough();

    jasmine.clock().tick(60000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(0);
    jasmine.clock().tick(60000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(0);
    jasmine.clock().tick(600000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(0);

    jasmine.clock().uninstall();
  });


  it('should clear timers when unmounted', () => {
    jasmine.clock().install();
    jasmine.clock().mockDate(new Date(2018, 7, 21, 1, 0, 0));
    const component = renderer.create(<TimedComponent prop="test" stops={stops}/>);

    spyOn(TimedComponent.prototype, 'render').and.callThrough();

    jasmine.clock().tick(60000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(1);
    component.unmount();
    jasmine.clock().tick(60000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(1);

    jasmine.clock().uninstall();
  });

  it('should clear timers when unmounted', () => {
    jasmine.clock().install();
    jasmine.clock().mockDate(new Date(2018, 7, 21, 1, 0, 20));
    const component = renderer.create(<TimedComponent prop="test" stops={stops}/>);

    spyOn(TimedComponent.prototype, 'render').and.callThrough();

    jasmine.clock().tick(40000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(1);
    jasmine.clock().tick(60000);
    expect(TimedComponent.prototype.render.calls.count()).toBe(2);

    jasmine.clock().uninstall();
  });
});