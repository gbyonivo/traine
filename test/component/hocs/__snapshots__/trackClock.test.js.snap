// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refresh should return a Refresh component when passed a named component 1`] = `
<div
  className="test"
/>
`;

exports[`refresh should return a TimedComponent component when passed a named component 1`] = `
<div
  className="test"
/>
`;

exports[`refresh should return a TimedComponent empty stops 1`] = `
<div
  className="test"
/>
`;

exports[`refresh should return a TimedComponent with empty stops 1`] = `
<div
  className="test"
/>
`;
